contracts/secret-storage/SecretStore.sol

```diff
@@ -0,0 +1,581 @@
+        delete userStorage[_secretId];
+
+        USDC_TOKEN.safeTransfer(owner, usdcDeposit);
+
+        emit EnclaveAcknowledgementFailed(_secretId);
+    }
+
+    function _markStoreAlive(
+        uint256 _secretId,
+        uint256 _signTimestamp,
+        bytes memory _signature,
+        address _owner
+    ) internal {
+        address enclaveAddress = _verifyStoreAliveSign(_secretId, _signTimestamp, _signature);
+
+        uint256 enclaveIndex = _getSelectedEnclaveIndex(_secretId, enclaveAddress);
+        if(!userStorage[_secretId].selectedEnclaves[enclaveIndex].hasAcknowledgedStore)
+            revert SecretStoreUnacknowledged();
+        if(block.timestamp > userStorage[_secretId].selectedEnclaves[enclaveIndex].lastAliveTimestamp + MARK_ALIVE_TIMEOUT)
+            revert SecretStoreMarkAliveTimeoutOver();
+
+        uint256 endTime;
+        if(block.timestamp > userStorage[_secretId].endTimestamp)
+            endTime = userStorage[_secretId].endTimestamp;
+        else
+            endTime = block.timestamp;
+
+        uint256 usdcPayment = ((endTime - userStorage[_secretId].selectedEnclaves[enclaveIndex].lastAliveTimestamp) * userStorage[_secretId].sizeLimit * STORAGE_FEE_RATE) / NO_OF_NODES_TO_SELECT;
+        userStorage[_secretId].usdcDeposit -= usdcPayment;
+        userStorage[_secretId].selectedEnclaves[enclaveIndex].lastAliveTimestamp = _signTimestamp;
+
+        USDC_TOKEN.safeTransfer(_owner, usdcPayment);
+
+        // TODO: delete from selectedNodes array and refund remaining usdc
+        if(block.timestamp > userStorage[_secretId].endTimestamp) {
+            _removeSelectedEnclave(_secretId, enclaveIndex);
+            if(userStorage[_secretId].selectedEnclaves.length == 0)
+                _refundExcessDepositAndRemoveStore(_secretId);
+        }
+
+        emit EnclaveStoreAlive(_secretId, enclaveAddress);
+    }
+
+    function _verifyStoreAliveSign(
+        uint256 _secretId,
+        uint256 _signTimestamp,
+        bytes memory _signature
+    ) internal view returns(address signer) {
+        _checkSignValidity(_signTimestamp);
+
+        bytes32 hashStruct = keccak256(abi.encode(ALIVE_TYPEHASH, _secretId, _signTimestamp));
+        bytes32 digest = keccak256(abi.encodePacked("\x19\x01", DOMAIN_SEPARATOR, hashStruct));
+        signer = digest.recover(_signature);
+
+        ENCLAVE_STORE.allowOnlyVerified(signer);
+    }
+
+    function _markStoreDead(
+        uint256 _secretId
+    ) internal {
+        for (uint256 index = 0; index < userStorage[_secretId].selectedEnclaves.length; ) {
+            bool isArrayLengthReduced = _markEnclaveDead(_secretId, userStorage[_secretId].selectedEnclaves[index].enclaveAddress);
+            if(!isArrayLengthReduced)
+                ++index;
+        }
+
+        // TODO: delete data and refund remaining
+        if(block.timestamp > userStorage[_secretId].endTimestamp && userStorage[_secretId].selectedEnclaves.length == 0)
```

Comment:
if endtimestamp has been reached, isn't it implicit that length of array of selected gateways would be zero. I mean only first condition is necessary, second is always true if first is.

Reply:
There can be a case where the enclaves are yet to send their alive check after the endTimestamp, and someone calls the function markStoreDead(). In this case, the selectedEnclaves length won't be zero.